lib/docker_container.rb

```diff
@@ -0,0 +1,119 @@
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+
+class DockerContainer
+  def initialize(image)
+    @instance = create_container(image)
+  end
+
+  def start
+    @instance.start!
+  end
+
+  def stop
+    @instance.delete(force: true)
+  end
+
+  def run_command(*args)
+    LoggedCheetah.run("docker", "exec", "-i", @container.id, *args)
+  end
+
+  # Copies a file to the container
+  def inject_file(source, destination)
+    LoggedCheetah.run("docker", "cp", source, "#{@container.id}:#{destination}")
+  end
+
+  # Retrieves files specified in filelist from the remote system and raises an
+  # Machinery::Errors::RsyncFailed exception when it's not successful. Destination is
+  # the directory where to put the files.
+  def retrieve_files(filelist, destination)
+    filelist.each do |file|
+      destination_dir = File.dirname(File.join(destination, file))
+
+      tar = StringIO.new
+      @container.copy(file) { |chunk| tar.write(chunk) }
+      tar.rewind
+
+      untar(tar, destination_dir)
+    end
+  end
+
+  # Retrieves files specified in file_list from the container and creates an archive.
+  def create_archive(file_list, archive, exclude = [])
+    created = !File.exists?(archive)
+    out = File.open(archive, "w")
+    begin
+      run_command(
+        File.join(
+          Machinery::REMOTE_HELPERS_PATH, "machinery-helper"
+        ), "tar", "--create", "--gzip", "--null", "--files-from=-",
+        *exclude.flat_map { |f| ["--exclude", f]},
+        stdout: out,
+        stdin: Array(file_list).join("\0"),
+        stderr: STDERR
+      )
+    rescue Cheetah::ExecutionFailed => e
+      if e.status.exitstatus == 1
+        # The tarball has been created successfully but some files were changed
+        # on disk while being archived, so we just log the warning and go on
+        Machinery.logger.info e.stderr
+      else
+        raise
+      end
+    end
+    out.close
+    File.chmod(0600, archive) if created
+  end
+
+  private
+
+  def untar(io, destination)
+    Gem::Package::TarReader.new io do |tar|
+      tar.each do |tarfile|
+        destination_file = File.join destination, tarfile.full_name
+
+        if tarfile.directory?
+          FileUtils.mkdir_p destination_file, mode: 0700
+        else
+          destination_directory = File.dirname(destination_file)
+          if !File.directory?(destination_directory)
+            FileUtils.mkdir_p(destination_directory, mode: 0700)
+          end
+
+          File.open destination_file, "wb" do |f|
+            f.print tarfile.read
+          end
+        end
+        File.chmod((tarfile.header.mode.to_s(8)[0..-3] + "00").to_i(8), destination_file)
+      end
+    end
+  end
+
+  def create_container(image)
+    validate_image_name(image)
+
+    @container = Docker::Container.create('Cmd' => ['bash'], "Tty" => true, "Image" => image)
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.